Completer Documentation Panel UI

ryantberg opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/master/CONTRIBUTING.md
-->

## References
A component of #7983 

This PR depends on #8080, and so will include some commits from that PR until it is merged.

The relevant changes actually proposed by this PR are [here](https://github.com/jupyterlab/jupyterlab/pull/8141/files/97adf0ed3dcadc10c8670c3aad4dec6cf6ed0728..8d6ec5ef8c042c774bf2587eb9a123597967e629)

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

<!-- Describe the code changes and how they address the issue. -->
#### Widget:
- Top-level `jp-Completer` is now a `<div>`, containing:
  - `jp-Completer-list`, the `<ul>` of completions
  - `jp-Completer-docpanel`, the `<div>` containing the docs for the active completion
    - this has a child `<pre>` that actually holds the docs text
- Added function `_updateDocPanel` which updates the inner contents of the docpanel and shows/hides it. It is called when the model is updated as well as when the user scrolls through the list.

#### CSS:
- New class `jp-Completer-list` for the new inner list
- New class `jp-Completer-docpanel` for the styling of the docpanel

## User-facing changes

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots here. -->

Some screenshots showing possible states:
Completion with no doc:
![Screen Shot 2020-03-18 at 4 27 05 PM](https://user-images.githubusercontent.com/12799462/77983930-12ef0700-72c5-11ea-84f6-8cec03d2f485.png)

Completion with a doc:
![Screen Shot 2020-03-27 at 2 30 49 PM](https://user-images.githubusercontent.com/12799462/77983885-f0f58480-72c4-11ea-82aa-06eb63cc3f71.png)

Completion with a long doc (docpanel is scrollable in both directions here, text does not wrap):
![Screen Shot 2020-03-27 at 2 35 09 PM](https://user-images.githubusercontent.com/12799462/77983898-f783fc00-72c4-11ea-9bd4-5e0c3fa0b1aa.png)

Dark theme:
![Screen Shot 2020-03-27 at 2 36 33 PM](https://user-images.githubusercontent.com/12799462/77983905-fce14680-72c4-11ea-8d69-30e4f9aa38cb.png)

Let me know if there are any other states you would like to see, and I will add them to the list above.

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->
It's possible that the CSS class changes represent a backwards-incompatible change, but I'm not sure. Can anyone shed some light on that?

tgeorgeux commented:
This is a great addition!

Can you match left padding and to the completion on the left? 

Would it make sense to add word-wrap instead of cutting off long lines?

ryantberg commented:
Agreed. Changed padding to match completions, and added text-wrapping to the docs.

![Screen Shot 2020-04-13 at 12 35 18 PM](https://user-images.githubusercontent.com/12799462/79167522-70428800-7d9c-11ea-9c5a-3cb660fe0b6a.png)

tgeorgeux commented:
Looks a lot better! I'm going to nitpick once more and ask for similar padding on all 4 sides. The way Lorem is butting up against the top in this makes it feel too cramped, maybe an 8px padding all around would be appropriate. 

Thanks for jumping on that!

telamonian commented on packages/completer/src/handler.ts:
```diff
@@ -477,6 +486,85 @@ export namespace CompletionHandler {
+     */
+    type?: string;
+
+    /**
+     * Image url for icon to be rendered with completion type.
+     */
+    icon?: string;
```
I have some questions about this:

- Where is this coming from, and where does it actually get used?
- Does it have to be implemented as an image url string?

Across the jlab codebase we use a consistent naming scheme for icon-related data in interfaces:

- `icon`: an instance of `LabIcon`, or of something that can be resolved to a `LabIcon`
- `iconClass`: the name of a class that has an associated CSS rule that sets some icon as the `background-image`

We don't take raw image urls as icons anywhere else, so ideally the above would get changed to eg `iconClass`. At the very least, this field needs to be called something other than `icon`

ryantberg replied:
> Where is this coming from, and where does it actually get used?

The purpose of the field is to display an icon given by completion providers next to items in the completer list. There's more information on this in #8140.

> Does it have to be implemented as an image url string?

SVG data-uris seemed like the simplest approach (easily serialized, doesn't require separate fetching/dereferencing).

However, given that extensions can provide styling and thereby add icons via custom `iconClass`es, that approach seems good as well. Providing icons that are generated on-the-fly would be much harder or maybe impossible, but I don't think that's a big use-case, and passing class names around instead of SVG data will make completion responses smaller. I will switch to using that method instead.

telamonian replied:
@saulshanabrook and I just had a discussion about this. For now, the interface for icons here should be just:

```typescript
  icon?: LabIcon
```

You would then implement the creation of the icon DOM node with:

```typescript
if (icon) {
  const iconNode = icon.element({className: 'jp-Completer-type'});
  li.appendChild(iconNode);
}
```

telamonian replied:
There's an argument to be made that there should be both an `icon` and an `iconClass` field here, but we can wait to hash that out until #8237 is resolved

ryantberg replied:
This approach makes a lot of sense to me, and I have made the changes here and in #8080
